### DISCUSSION TITLE
React Query v3 Odd Behavior in Test Environment

### DISCUSSION THREAD

**Discussion by curtvict** [TanStack/query#6065] [General]
I'm working to standardize our query keys to use simple factories, sort of based around [this post](https://tkdodo.eu/blog/effective-react-query-keys). Here's what one looks like:

```ts
export const partnerQueryKeys = {
  getPartners: (params?: GetPartnersParams) => ['getPartners', params] as const,
};
```

Introducing it into the codebase looked like this:

![image](https://github.com/TanStack/query/assets/96080054/776e082c-de19-4396-a89b-4f3ba72fe485)

The factory supplies the query key, in array format, adding the `{ limit: 1 }` param bit to the key. For simplicity's sake I pull the `params` object back out of the `queryKey` object on the next line. 

Easy. TypeScript loves it. Works in the browser.

However in the Jest environment I'm receiving a `Missing queryFn` error. The weird part is that when switch the above code in the screenshot to this, the tests stop complaining: 

```tsx
const fetchingPartner = useQuery({
  queryKey: partnerQueryKeys.getPartners({ limit: 1 }),
  queryFn: ({ queryKey: [, params] }) => getPartners(params)
});
```

It's odd since the signature (`useQuery(queryKey, queryFn, options)`) worked before the introduction of the query key generator but now seems to lose the query function somewhere in the mix.

Does anyone have any idea what might be going wrong at the proverbial post office?

**Comment by TkDodo** [comment]
weird, but fyi, the object syntax is the only one that will be supported from v5 onwards, so it's a good idea to switch to it regardless :)

**Reply by curtvict** [reply]
Ah, yes, I remember reading that in the upgrade guide to 4, I believe. I've been trying to incrementally make consistent our usage so we can finally bump to 4, so this will be helpful in the long run. Thanks for the reminder, and thanks for the great library!
